File: Array/Java/_2053_KthDistinctStringInAnArray.java
package Array.Java;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class _2053_KthDistinctStringInAnArray {

  public static void main(String[] args) throws Exception {
  }

  public String kthDistinctV1(String[] arr, int k) {
    List<String> all = new ArrayList<String>();
    List<String> dist = new ArrayList<String>();
    String str = "";

    for (int i = 0; i < arr.length; i++) {
        if (dist.contains(arr[i]) && all.contains(arr[i])) {
            dist.remove(arr[i]);
        }
        else if (dist.contains(arr[i]) == false && all.contains(arr[i])) {
            continue;
        }
        else {
            dist.add(arr[i]);
            all.add(arr[i]);
        }
    }

    if (dist.size() >= k) {
        str = dist.get(k - 1);
    }

    return str;
  }

  /**
   * Iteration
   * Time Complexity: BigO(N)
   * Space Complexity: BigO(1)
   */
  public String kthDistinctV2(String[] arr, int k) {
    ArrayList<String> v = new ArrayList<>();

    for(String s:arr)
        v.add(s);

    for(String s : v) {
        if(Collections.frequency(v, s) == 1) 
            k--;

        if(k == 0)
            return s;   
    }
    
    return "";
  }
}

File: Array/Python/_2053_KthDistinctStringInAnArray.py
from typing import List
from collections import Counter


class Solution:
    # Iteration
    # Time Complexity: BigO(N)
    # Space Complexity: BigO(N)
    def kthDistinctV1(self, arr: List[str], k: int) -> str:
        all_list = []
        dist_list = []
        for item in arr:
            if item not in dist_list and item not in all_list:
                dist_list.append(item)
                all_list.append(item)
            elif item not in dist_list and item in all_list:
                continue
            else:
                dist_list.remove(item)

        if len(dist_list) < k:
            return ""

        return dist_list[k - 1]

    # Iteration
    # Time Complexity: BigO(N)
    # Space Complexity: BigO(N)
    def kthDistinctV2(self, arr: List[str], k: int) -> str:
        dist_list = Counter(arr)

        for str in dist_list:
            if dist_list[str] == 1:
                k -= 1
            if k == 0:
                return str

        return ""
